File: examples/soc_simple.lua
local R = require "rigel"
local SOC = require "generators.soc"
local C = require "generators.examplescommon"
local harness = require "generators.harnessSOC"
local G = require "generators.core"
local RS = require "rigelSimple"
local types = require "types"
types.export()
local SDF = require "sdf"
local Zynq = require "generators.zynq"

R.MONITOR_FIFOS = true

local regs = SOC.axiRegs({},SDF{1,1024}):instantiate("regs")

local noc = Zynq.SimpleNOC(nil,nil,{{regs.read,regs.write}} ):instantiate("ZynqNOC")
noc.extern=true

local OffsetModule = G.Function{ "OffsetModule", R.HandshakeTrigger,
  function(i)
    local readStream = G.AXIReadBurst{ "frame_128.raw", {128,64}, u(8), 8, noc.read }(i)
    local offset = G.Map{G.Add{200}}(readStream)
    return G.AXIWriteBurst{"out/soc_simple",noc.write}(offset)
  end}

harness({regs.start, OffsetModule, regs.done},nil,{regs})

File: cppexamples/soc_simple.cpp
#include <fstream>
#include <rigel.h>
#include <types.h>
#include <library.h>
#include <string>

using namespace Rigel;
using namespace RigelLibrary;

class Add200:public UserFunction{
public:
  Add200():UserFunction("Add200",Trigger()){}
  Val define(Val inp){
    Val readStream = AXIReadBurst<128,64,8>(inp);
    Val offset = Map< Add<200> >(readStream);
    return AXIWriteBurst(offset);
  }
};


int main(){
  Add200 add200;
  std::string add200lua = add200.luaDefinition();

  std::ofstream out("out/soc_simple.lua");
  out << header;
  out << "local regs = SOC.axiRegs({},SDF{1,1024}):instantiate(\"regs\")\n"
"local noc = Zynq.SimpleNOC(nil,nil,{{regs.read,regs.write}} ):instantiate(\"ZynqNOC\")\n"
"noc.extern=true\n";
  out << add200lua;
  out << footer( add200, "soc_simple" );
  out.close();

  return 0;
}
